src/town_cmd.cpp

```diff
@@ -3132,8 +3135,88 @@ static void UpdateTownRating(Town *t)
 	SetWindowDirty(WC_TOWN_AUTHORITY, t->index);
 }
 
-static void UpdateTownGrowRate(Town *t)
+
+/**
+ * Updates town grow counter after growth rate change.
+ * Preserves relative house builting progress whenever it can.
+ * @param town The town to calculate grow counter for
+ * @param prev_growth_rate Town growth rate before it changed (one that was used with grow counter to be updated)
+ */
+static void UpdateTownGrowCounter(Town *t, uint16 prev_growth_rate)
+{
+	if (t->growth_rate == TOWN_GROWTH_RATE_NONE) return;
+	if (prev_growth_rate == TOWN_GROWTH_RATE_NONE) {
+		t->grow_counter = min(t->growth_rate, t->grow_counter);
+		return;
+	}
+	t->grow_counter = RoundDivSU((uint32)t->grow_counter * (t->growth_rate + 1), prev_growth_rate + 1);
+}
+
+/**
+ * Calculates amount of active stations in the range of town (HZB_TOWN_EDGE).
+ * @param town The town to calculate stations for
+ * @returns Amount of active stations
+ */
+static int CountActiveStations(Town *t)
+{
+	int n = 0;
+	const Station *st;
+	FOR_ALL_STATIONS(st) {
+		if (DistanceSquare(st->xy, t->xy) <= t->cache.squared_town_zone_radius[0]) {
+			if (st->time_since_load <= 20 || st->time_since_unload <= 20) {
+				n++;
+			}
+		}
+	}
+	return n;
+}
+
+/**
+ * Calculates town growth rate in normal conditions (custom growth rate not set).
+ * @param town The town to calculate growth rate for
+ * @returns Calculated growth rate
+ */
+static uint GetNormalGrowthRate(Town *t)
 {
+	static const uint16 _grow_count_values[2][6] = {
+		{ 120, 120, 120, 100,  80,  60 }, // Fund new buildings has been activated
+		{ 320, 420, 300, 220, 160, 100 }  // Normal values
+	};
+
+	int n = CountActiveStations(t);
+	uint16 m = _grow_count_values[t->fund_buildings_months != 0 ? 0 : 1][min(n, 5)];
+
+	/* Use the normal growth rate values if new buildings have been funded in
+	 * this town and the growth rate is set to none. */
+	uint growth_multiplier = _settings_game.economy.town_growth_rate != 0 ? _settings_game.economy.town_growth_rate - 1 : 1;
```

Comment:
The comment for this line is really confusing on its own. I realize this is just moved code, but I think it'd be a good idea to mention that the second part of this logic happens in `UpdateTownGrowth` , which aborts on the condition of `economy.town_growth_rate == 0 && fund_buildings_months == 0`.

Reply:
I agree that comment is a bit confusing, but it's just commenting the line below it, nothing more. Since I haven't modified that line I left the comment as well. Again, improving existing code is not a goal of this patch.

Reply:
How I see it, you made the comment worse by moving code around, and part of the code the comment applies to is now in a different function. If you don't keep comments updated when you refactor code, even though none of the logic the comment is explaining has changed, you're still contributing to "comment rot", i.e. the observation that comments in code tend to become incorrect over time as code changes around them.
(I see you just now pushed an update that expands the function comment, at least partially addressing this.)

Reply:
After a bit of thought I decided it's better to be mentioned in function description rather than regular comment.

Reply:
@nielsmh Didn't see your reply yesterday. You seem to really misunderstand that comment, it has nothing to do with other functions, it pretty much repeats the line below. Growth rate for speed "None" is the same as for "Normal". If you follow the rest of the code then yes, probably the only case where it matters is when buildings are funded in "None" speed. But it has nothing to do with this function, functions don't decide how they're called. Anyway, it's already changed.